### ISSUE TITLE
chown privileges on OpenShift

### ISSUE THREAD

**Issue by jkogler-cloudflight** [pypiserver/pypiserver#357]
I'm running an pypiserver instance on our OpenShift platform with a volume mounted at `/data`.
On OpenShift every container is started with a random non-privileged user (even if there is a user defined in the Dockerfile). This means that the command 
https://github.com/pypiserver/pypiserver/blob/15d81147ac51b558ea5a885eac900434d387b04d/entrypoint.sh#L9
will fail, since the non-privileged user has no permissions to change ownership of files to other users. And therefore the server doesn't start and fail with `chown: Operation not permitted`.

The same was also true in the previous version where you chowned the entire `/data` directory.

The last version that worked was before PR #330, which introduced the `entrypoint.sh` script with the chown.

As a workaround I currently simply overwrite the entrypoint of the container with `pypi-server -p 8080 -P . -a . packages` to ignore the `entrypoint.sh` and everything works fine. Although that's not a clean solution.

---

If I understand https://docs.openshift.com/container-platform/3.11/creating_images/guidelines.html#openshift-specific-guidelines correctly, reading and writing works on OpenShift Volumes without any `chown`, because the random user is part of the `root` group and that group has read/write privileges on the volumes (and also on the image itself if not chowned to a different user/group).

---

From what I've seen in the history of PR #330, that `chown` command was actually introduced because it was required for local `docker-compose` permissions. So removing it would probably not work.
Any idea how other projects handle that issue?

**Comment by elfjes** [comment]
Thanks for your bug report. I agree with you that the way the docker image is set up now limits the number of configurations in which it can be used. Maybe instead of chowning in the entrypoint, we can chown in the dockerfile, as layed out here: https://github.com/docker/compose/issues/3270. Something like this:
```Dockerfile
RUN mkdir -p /data/packages && chown pypiserver:pypiserver /data/packages
VOLUME /data/packages
USER pypiserver
```
The volume would still be chowned for docker compose, but we don't have to use a privileged user at runtime.

**Comment by mplanchard** [comment]
Yeah I believe we used to do it that way (changed in #330), but this led to confusion about permissions also, due to docker's different permissions model on mac vs linux systems (see #256).

The created user has a predictable user/group ID (both are 9898), but it seems like OpenShift's approach may make it difficult to run as a user with that ID. You might could use the second option that I recommended in #256 (run with `--user $(id -u $(whoami))`.

I do agree we can make some improvements to the dockerfile. In retrospect, the changes in #330 have caused more problems than they've solved IMO, and I probably shouldn't have accepted the patch.

**Comment by elfjes** [comment]
So I guess we can identify a number of requirements:
- works with docker compose
- works when it doesn't need to / cannot `chown` packages (ie. openshift)
- works on both linux and mac
- works with existing packages that may have different permissions? We may choose not to have this requirement I guess, since it's an operations misconfiguration imho.

anything else?

**Comment by mplanchard** [comment]
Those requirements look good to me, other than the last one, which I agree with you is largely beyond our control.

I'd also add as an explicit requirement:

* doesn't break current configs if at all possible

As I think I've mentioned before, when I first added the Dockerfile, I didn't know whether or not it would be a commonly used way of running pypiserver, but it seems to be one of the most common ways that people run it now. So, we definitely need to improve the testing on it and also try to make sure any changes we make don't break people's setups.

Wanting to note on this requirement:

> * works when it doesn't need to / cannot chown packages (ie. openshift)

That it is a bit of a tough requirement, but I think that if we keep the entrypoint.sh script, and update it to only attempt a chown if the pypiserver doesn't have read/write access to the directory, we might be good.

**Comment by elfjes** [comment]
> Those requirements look good to me, other than the last one, which I agree with you is largely beyond our control.
> 
> I'd also add as an explicit requirement:
> 
>     * doesn't break current configs if at all possible
I've added that to the list
> 
> 
> As I think I've mentioned before, when I first added the Dockerfile, I didn't know whether or not it would be a commonly used way of running pypiserver, but it seems to be one of the most common ways that people run it now. So, we definitely need to improve the testing on it and also try to make sure any changes we make don't break people's setups.
> 
> Wanting to note on this requirement:
> 
> > * works when it doesn't need to / cannot chown packages (ie. openshift)
> 
> That it is a bit of a tough requirement, but I think that if we keep the entrypoint.sh script, and update it to only attempt a chown if the pypiserver doesn't have read/write access to the directory, we might be good.

Another possible solution might be to just try chowning and not crash when it fails (but log a meaningful message though)

**Comment by mplanchard** [comment]
We've just merged some updates to the docker image into the master branch that I hope will address these issues. If you've got time, could you check out the `unstable` branch at some point and let us know if it works for you?

**Comment by arielshulman** [comment]
We've tried the `unstable` image on our openshift environment and it worked fine after one tweek.
The only weird thing was that the packages folder was mount on `/data/packages` as it was before, and no package was shown.
After searching a bit in the code, I've tried to mount it to `/packages` and it worked only the script kept searching for `/data/packages` so I've mount the folder twice for `/packages` and `/data/packages`  🤷‍♂️
